Truncating an MFS file 175 times makes it unwritable.

After exactly 175 truncate/writes, mfs returns an empty string from read (forever, apparently).

@Stebalien Are you currently addressing this in your PR #4517? (I saw you were making fixes to the [`Truncate`](https://github.com/ipfs/go-ipfs/pull/4517/files#diff-8025428853309129ba9ae4df8ea67f9e) function). If not I would like to give this a try, seems like an interesting issue.

Truncating to any smaller than (the original) 4 bytes value produces the same error. Truncating to zero 174 times and the rest of the times to 1 produces:

```
--- FAIL: TestTruncateAndWrite (0.20s)
	mfs_test.go:1168: read error at read 175, read: [116 116 101 115 116]
	mfs_test.go:1168: read error at read 176, read: [116]
	mfs_test.go:1168: read error at read 177, read: [116]
	[...]
```

Similarly, truncating to 2:

```
--- FAIL: TestTruncateAndWrite (0.20s)
	mfs_test.go:1168: read error at read 175, read: [116 101 116 101 115 116]
	mfs_test.go:1168: read error at read 176, read: [116 101]
	mfs_test.go:1168: read error at read 177, read: [116 101]
```

The same progression continues truncating to 3. I'm avoiding truncating to the same size for the error you encountered and fixed. Truncating to 5 or bigger (always starting with 174 truncations to zero) only fails in iteration 175:

```
--- FAIL: TestTruncateAndWrite (0.22s)
	mfs_test.go:1168: read error at read 175, read: [116 101 115 116 0 116 101 115 116]
FAIL
FAIL	github.com/ipfs/go-ipfs/mfs	0.222s
```

Truncating from the start to a value bigger than 4 panics:

```
--- FAIL: TestTruncateAndWrite (0.00s)
	mfs_test.go:1147: read error at read 0, read: [0 0]
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x6ac1c6]

goroutine 22 [running]:
testing.tRunner.func1(0xc4201263c0)
	/home/user/go/src/github.com/golang/go/src/testing/testing.go:711 +0x2d2
panic(0x73b180, 0x97e4b0)
	/home/user/go/src/github.com/golang/go/src/runtime/panic.go:491 +0x283
github.com/ipfs/go-ipfs/blockservice.(*blockService).AddBlock(0xc42007f4a0, 0x0, 0x0, 0x0, 0x0)
	/home/user/go/src/github.com/ipfs/go-ipfs/blockservice/blockservice.go:132 +0x26
github.com/ipfs/go-ipfs/merkledag.(*dagService).Add(0xc42007b250, 0x953940, 0xc420086058, 0x0, 0x0, 0xc4200871a8, 0x0)
	/home/user/go/src/github.com/ipfs/go-ipfs/merkledag/merkledag.go:52 +0x8c
github.com/ipfs/go-ipfs/unixfs/mod.dagTruncate(0x953940, 0xc420086058, 0x956080, 0xc4201255e0, 0x6, 0x954e40, 0xc42007b250, 0x0, 0x40, 0x40, ...)
	/home/user/go/src/github.com/ipfs/go-ipfs/unixfs/mod/dagmodifier.go:573 +0x448
github.com/ipfs/go-ipfs/unixfs/mod.(*DagModifier).Truncate(0xc420138120, 0x6, 0x21, 0x2)
	/home/user/go/src/github.com/ipfs/go-ipfs/unixfs/mod/dagmodifier.go:505 +0xc5
github.com/ipfs/go-ipfs/mfs.(*fileDescriptor).Truncate(0xc42010a5c0, 0x6, 0x1f, 0xc42004df78)
	/home/user/go/src/github.com/ipfs/go-ipfs/mfs/fd.go:49 +0x9c
github.com/ipfs/go-ipfs/mfs.TestTruncateAndWrite(0xc4201263c0)
	/home/user/go/src/github.com/ipfs/go-ipfs/mfs/mfs_test.go:1131 +0x38a
testing.tRunner(0xc4201263c0, 0x7b7500)
	/home/user/go/src/github.com/golang/go/src/testing/testing.go:746 +0xd0
created by testing.(*T).Run
	/home/user/go/src/github.com/golang/go/src/testing/testing.go:789 +0x2de
```

But the last two cases may be more related with a problem in `expandSparse` than the issue in  `dagTruncate` that is causing the error you reported.

--- a/mfs/mfs_test.go
+++ b/mfs/mfs_test.go
@@ -1136,3 +1136,43 @@ func TestTruncateAtSize(t *testing.T) {
 	}
 	fd.Truncate(4)
 }
+
+func TestTruncateAndWrite(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ds, rt := setupRoot(ctx, t)
+
+	dir := rt.GetDirectory()
+
+	nd := dag.NodeWithData(ft.FilePBData(nil, 0))
+	fi, err := NewFile("test", nd, dir, ds)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	fd, err := fi.Open(OpenReadWrite, true)
+	defer fd.Close()
+	if err != nil {
+		t.Fatal(err)
+	}
+	for i := 0; i < 200; i++ {
+		err = fd.Truncate(0)
+		if err != nil {
+			t.Fatal(err)
+		}
+		l, err := fd.Write([]byte("test"))
+		if err != nil {
+			t.Fatal(err)
+		}
+		if l != len("test") {
+			t.Fatal("incorrect write length")
+		}
+		data, err := ioutil.ReadAll(fd)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if string(data) != "test" {
+			t.Errorf("read error at read %d, read: %v", i, data)
+		}
+	}
+}

--- a/test/sharness/t0250-files-api.sh
+++ b/test/sharness/t0250-files-api.sh
@@ -613,7 +613,7 @@ tests_for_files_api() {
   ROOT_HASH=QmcwKfTMCT7AaeiD92hWjnZn9b6eh9NxnhfSzN5x2vnDpt
   CATS_HASH=Qma88m8ErTGkZHbBWGqy1C7VmEmX8wwNDWNpGyCaNmEgwC
   FILE_HASH=QmQdQt9qooenjeaNhiKHF3hBvmNteB4MQBtgu3jxgf9c7i
-  TRUNC_HASH=QmdaQZbLwK5ykweGdCVovNnvBom7QhikovDUVqTPHQG4L8
+  TRUNC_HASH=QmPVnT9gocPbqzN4G6SMp8vAPyzcjDbUJrNdKgzQquuDg4
   test_files_api "($EXTRA)"
 
   test_expect_success "can create some files for testing with raw-leaves ($EXTRA)" '
@@ -629,13 +629,13 @@ tests_for_files_api() {
   ROOT_HASH=QmW3dMSU6VNd1mEdpk9S3ZYRuR1YwwoXjGaZhkyK6ru9YU
   CATS_HASH=QmPqWDEg7NoWRX8Y4vvYjZtmdg5umbfsTQ9zwNr12JoLmt
   FILE_HASH=QmRCgHeoKxCqK2Es6M6nPUDVWz19yNQPnsXGsXeuTkSKpN
-  TRUNC_HASH=QmRFJEKWF5A5FyFYZgNhusLw2UziW9zBKYr4huyHjzcB6o
+  TRUNC_HASH=QmckstrVxJuecVD1FHUiURJiU9aPURZWJieeBVHJPACj8L
   test_files_api "($EXTRA, raw-leaves)" '' --raw-leaves
 
   ROOT_HASH=QmageRWxC7wWjPv5p36NeAgBAiFdBHaNfxAehBSwzNech2
   CATS_HASH=zdj7WkEzPLNAr5TYJSQC8CFcBjLvWFfGdx6kaBrJXnBguwWeX
   FILE_HASH=zdj7WYHvf5sBRgSBjYnq64QFr449CCbgupXfBvoYL3aHC1DzJ
-  TRUNC_HASH=zdj7WYLYbka6Ydg8gZUJRLKnFBVehCADhQKBsFbNiMxZSB5Gj
+  TRUNC_HASH=zdj7Wjr8GHZonPFVCWvz2SLLo9H6MmqBxyeB34ArHfyCbmdJG
   if [ "$EXTRA" = "offline" ]; then
     test_files_api "($EXTRA, cidv1)" --cid-version=1
   fi
@@ -660,7 +660,7 @@ tests_for_files_api() {
     ROOT_HASH=zDMZof1kxEsAwSgCZsGQRVcHCMtHLjkUQoiZUbZ87erpPQJGUeW8
     CATS_HASH=zDMZof1kuAhr3zBkxq48V7o9HJZCTVyu1Wd9wnZtVcPJLW8xnGft
     FILE_HASH=zDMZof1kxbB9CvxgRioBzESbGnZUxtSCsZ18H1EUkxDdWt1DYEkK
-    TRUNC_HASH=zDMZof1kxXqKdVsVo231qVdN3hCTF5a34UuQZpzmm5K7CbRJ4u2S
+    TRUNC_HASH=zDMZof1kpH1vxK3k2TeYc8w59atCbzMzrhZonsztMWSptVro2zQa
     test_files_api "($EXTRA, blake2b-256 root)"
   fi
 

--- a/unixfs/mod/dagmodifier.go
+++ b/unixfs/mod/dagmodifier.go
@@ -529,7 +529,13 @@ func dagTruncate(ctx context.Context, n ipld.Node, size uint64, ds ipld.DAGServi
 	var cur uint64
 	end := 0
 	var modified ipld.Node
-	ndata := ft.NewFSNode(ft.TRaw)
+	ndata, err := ft.FSNodeFromBytes(nd.Data())
+	if err != nil {
+		return nil, err
+	}
+	// Reset the block sizes of the node to adjust them
+	// with the new values of the truncated children.
+	ndata.RemoveAllBlockSizes()
 	for i, lnk := range nd.Links() {
 		child, err := lnk.GetNode(ctx, ds)
 		if err != nil {
@@ -558,7 +564,7 @@ func dagTruncate(ctx context.Context, n ipld.Node, size uint64, ds ipld.DAGServi
 		ndata.AddBlockSize(childsize)
 	}
 
-	err := ds.Add(ctx, modified)
+	err = ds.Add(ctx, modified)
 	if err != nil {
 		return nil, err
 	}
@@ -573,7 +579,7 @@ func dagTruncate(ctx context.Context, n ipld.Node, size uint64, ds ipld.DAGServi
 	if err != nil {
 		return nil, err
 	}
-
+	// Save the new block sizes to the original node.
 	nd.SetData(d)
 
 	// invalidate cache and recompute serialized data

--- a/unixfs/mod/dagmodifier_test.go
+++ b/unixfs/mod/dagmodifier_test.go
@@ -406,6 +406,44 @@ func testDagTruncate(t *testing.T, opts testu.NodeOpts) {
 	}
 }
 
+// TestDagTruncateSameSize tests that a DAG truncated
+// to the same size (i.e., doing nothing) doesn't modify
+// the DAG (its hash).
+func TestDagTruncateSameSize(t *testing.T) {
+	runAllSubtests(t, testDagTruncateSameSize)
+}
+func testDagTruncateSameSize(t *testing.T, opts testu.NodeOpts) {
+	dserv := testu.GetDAGServ()
+	_, n := testu.GetRandomNode(t, dserv, 50000, opts)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	dagmod, err := NewDagModifier(ctx, n, dserv, testu.SizeSplitterGen(512))
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Copied from `TestDagTruncate`.
+
+	size, err := dagmod.Size()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = dagmod.Truncate(size)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	modifiedNode, err := dagmod.GetNode()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if modifiedNode.Cid().Equals(n.Cid()) == false {
+		t.Fatal("the node has been modified!")
+	}
+}
+
 func TestSparseWrite(t *testing.T) {
 	runAllSubtests(t, testSparseWrite)
 }

--- a/unixfs/unixfs.go
+++ b/unixfs/unixfs.go
@@ -201,6 +201,12 @@ func (n *FSNode) BlockSize(i int) uint64 {
 	return n.format.Blocksizes[i]
 }
 
+// RemoveAllBlockSizes removes all the child block sizes of this node.
+func (n *FSNode) RemoveAllBlockSizes() {
+	n.format.Blocksizes = []uint64{}
+	n.format.Filesize = proto.Uint64(uint64(len(n.Data())))
+}
+
 // GetBytes marshals this node as a protobuf message.
 func (n *FSNode) GetBytes() ([]byte, error) {
 	return proto.Marshal(&n.format)

